OSDOCS-9565: configuring AWS Outposts postinstallation (#71263)
openshift/openshift-docs: modules/nw-aws-load-balancer-with-outposts.adoc

Reviewer: Posted a review: https://github.com/openshift/openshift-docs/pull/71263#pullrequestreview-1894479668.

// Module included in the following assemblies:
// * networking/aws_load_balancer_operator/understanding-aws-load-balancer-operator.adoc
// * post_installation_configuration/configuring-aws-outposts.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-aws-load-balancer-with-outposts_{context}"]
= Using the AWS Load Balancer Operator in an AWS VPC cluster extended into an Outpost

> You can configure the AWS Load Balancer Operator to provision an AWS Application Load Balancer in an AWS VPC cluster extended into an Outpost.
> AWS Outposts does not support AWS Network Load Balancers.
> As a result, the AWS Load Balancer Operator cannot provision Network Load Balancers in an Outpost.
> 
> You can create an AWS Application Load Balancer either in the cloud subnet or in the Outpost subnet.
> An Application Load Balancer in the cloud can attach to cloud-based compute nodes and an Application Load Balancer in the Outpost can attach to edge compute nodes.
> You must annotate Ingress resources with the Outpost subnet or the VPC subnet, but not both.

Reviewer 2: This all seems like important info, but I found it hard to follow. Suggestion to take or leave for lightly restructuring this intro:

```suggestion
When you configure the AWS Load Balancer Operator to provision an AWS Application Load Balancer in an AWS VPC cluster extended into an Outpost, you must ensure that you annotate the `Ingress` resources with either the Outpost subnet or the VPC subnet, but not both.

Consider the following points when selecting a subnet:

* AWS Network Load Balancers are not supported by AWS Outposts. As a result, the AWS Load Balancer Operator cannot provision Network Load Balancers in an Outpost.
* You can create an AWS Application Load Balancer in the cloud subnet, where it can attach to cloud-based compute nodes, or in the Outpost subnet, where it can attach to edge compute nodes.
```

Also, not part of your PR, but the step and the config block title both need backticks around `Ingress`

// Module included in the following assemblies:
// * networking/aws_load_balancer_operator/understanding-aws-load-balancer-operator.adoc
// * post_installation_configuration/configuring-aws-outposts.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-aws-load-balancer-with-outposts_{context}"]

> = Using the AWS Load Balancer Operator with AWS Outposts

Reviewer: ```suggestion
= Using the AWS Load Balancer Operator in an AWS VPC cluster extended into an Outpost
```

> You can configure the AWS Load Balancer Operator to provision an AWS Application Load Balancer in an Outpost.

Reviewer: ```suggestion
You can configure the AWS Load Balancer Operator to provision an AWS Application Load Balancer in an AWS VPC cluster extended to an Outpost.
```

AWS Outposts does not support AWS Network Load Balancers or Classic Load Balancers.

> As a result, the AWS Load Balancer Operator cannot provision those types of load balancers in an Outpost.

Author: @alebedev87 is this statement needed? 
"As a result, the AWS Load Balancer Operator cannot provision those types of load balancers in an Outpost."

Per @lihongan, [CLB doesn't need AWS LB Operator](https://github.com/openshift/openshift-docs/pull/71263#pullrequestreview-1892494822), so I'm trying to reconcile these two statements.

Reviewer: You can create NLBs (like CCM, in that context they overlap in functionality) with ALBO. So we need to state in the docs that ALBO won't be able to provision NLBs. However CLBs are not used by ALBO, so they can be removed from ALBO's chapter.

Author: Aha - that all makes sense now. Thanks!

> You can create an individual AWS Application Load Balancer either in the cloud subnet or in the Outpost subnet, but they must be separate instances.

Reviewer: I'm not sure I understand the meaning of "individual" and "must be separate instances" in this context. May be make this phrase simpler:
```suggestion
You can create an AWS Application Load Balancer either in the cloud subnet or in the Outpost subnet.
```

An Application Load Balancer in the cloud can attach to cloud-based compute nodes and an Application Load Balancer in the Outpost can attach to edge compute nodes.

> You must annotate Ingress resources with the Outpost subnet or the VPC subnet, but not both.

Reviewer: I forgot that there must be at least 2 subnets specified for an ALB:
```suggestion
You must annotate Ingress resources with the Outpost subnets or the VPC subnets, but not both.
```

.Prerequisites

* You have extended an AWS VPC cluster into an Outpost.

> * You have access to the cluster using an account with `cluster-admin` permissions.

Reviewer: Why `cluster-admin` permissions? It's not needed for creating an `Ingress`.

Author: That was just an assumption on my part, based on many other commands. Will remove it, thanks!

* You have installed the {oc-first}.

* You have installed the AWS Load Balancer Operator and created the AWS Load Balancer Controller.

> .Procedure

Reviewer 3: I think we should add the NodePort service example before creating the ingress resource, because this is main difference from Classic Load Balancer/Network Load Balancer. 
Application Load Balancer needs both NodePort serivice and Ingress resource.
 
```
.Example `service` resource
[source,yaml]
----
apiVersion: v1
kind: Service # <1>
metadata:
  name:  <application_name>
  namespace: <application_namespace>
spec:
  ports:
    - port: 80
      targetPort: 8080
      protocol: TCP
  type: NodePort
  selector: # <2>
    app: <application_name>
```

Author: Thanks @lihongan. 

The intent was to show the config difference required for the ALB in this environment. It looks like the NodePort step is part of the task "[Creating an instance of the AWS Load Balancer Controller using AWS Load Balancer Operator](https://71263--ocpdocs-pr.netlify.app/openshift-enterprise/latest/networking/aws_load_balancer_operator/create-instance-aws-load-balancer-controller#nw-creating-instance-aws-load-balancer-controller_create-instance-aws-load-balancer)". Right now, I have the additional resources link pointing to general AWS LB Operator content, but I think I will update it to point to this specific task to help the user better. 

It might be a good enhancement later to simply pull that entire task into this age, with some conditions for the Outpost-specific details.

Cc: @alebedev87 and @jmanthei for awareness

apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: <application_name>
  annotations:

>     alb.ingress.kubernetes.io/subnets: <subnet_id> # <1>

Reviewer: There mush be at least 2 subnets from different availability zones.

Reviewer 4: Perhaps there are some exception cases?

```
$ oc get ingress alb-outpost-public -n ${APP_NAME} -o yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    alb.ingress.kubernetes.io/scheme: internet-facing
    alb.ingress.kubernetes.io/subnets: subnet-007223c427204d300
    alb.ingress.kubernetes.io/target-node-labels: location=outposts
    alb.ingress.kubernetes.io/target-type: instance
  creationTimestamp: "2024-02-22T06:12:22Z"
  finalizers:
  - ingress.k8s.aws/resources
  generation: 1
  labels:
    location: outpost
  name: alb-outpost-public
  namespace: myapp-outpost
  resourceVersion: "42950"
  uid: a38cd630-07d0-4e0a-bd05-cdf3c3d9ef8f
spec:
  ingressClassName: cloud
  rules:
  - http:
      paths:
      - backend:
          service:
            name: myapp-outpost-alb
            port:
              number: 80
        path: /
        pathType: Prefix
status:
  loadBalancer:
    ingress:
    - hostname: k8s-myappout-alboutpo-54dc972b7a-1655616622.us-east-1.elb.amazonaws.com
```

![Screenshot from 2024-02-22 03-16-52](https://github.com/openshift/openshift-docs/assets/3216894/7aff2b78-3937-48bb-93fa-2ecd5737ede7)
![Screenshot from 2024-02-22 03-17-26](https://github.com/openshift/openshift-docs/assets/3216894/c9cfaa89-18fc-4f1e-927c-4b9198916858)

Author: The way this is phrased, it could be applied to either. I can add context that the non-Outpost case needs at least two in different AZs so it's completely clear.
